controllers/kustomization_controller.go

```diff
@@ -127,279 +126,291 @@ func (r *KustomizationReconciler) SetupWithManager(mgr ctrl.Manager, opts Kustom
-
-	// Add our finalizer if it does not exist
-	if !controllerutil.ContainsFinalizer(&kustomization, kustomizev1.KustomizationFinalizer) {
-		controllerutil.AddFinalizer(&kustomization, kustomizev1.KustomizationFinalizer)
-		if err := r.Update(ctx, &kustomization); err != nil {
-			log.Error(err, "unable to register finalizer")
-			return ctrl.Result{}, err
-		}
-	}
-
-	// Examine if the object is under deletion
-	if !kustomization.ObjectMeta.DeletionTimestamp.IsZero() {
-		return r.reconcileDelete(ctx, kustomization)
-	}
+	defer r.RecordSuspend(ctx, kustomization, kustomization.Spec.Suspend)
 
 	// Return early if the Kustomization is suspended.
 	if kustomization.Spec.Suspend {
 		log.Info("Reconciliation is suspended for this object")
 		return ctrl.Result{}, nil
 	}
 
-	// resolve source reference
-	source, err := r.getSource(ctx, kustomization)
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(kustomization, r.Client)
 	if err != nil {
-		if apierrors.IsNotFound(err) {
-			msg := fmt.Sprintf("Source '%s' not found", kustomization.Spec.SourceRef.String())
-			kustomization = kustomizev1.KustomizationNotReady(kustomization, "", kustomizev1.ArtifactFailedReason, msg)
-			if err := r.patchStatus(ctx, req, kustomization.Status); err != nil {
-				log.Error(err, "unable to update status for source not found")
-				return ctrl.Result{Requeue: true}, err
-			}
-			r.recordReadiness(ctx, kustomization)
-			log.Info(msg)
-			// do not requeue immediately, when the source is created the watcher should trigger a reconciliation
-			return ctrl.Result{RequeueAfter: kustomization.GetRetryInterval()}, nil
-		} else {
-			// retry on transient errors
-			return ctrl.Result{Requeue: true}, err
-		}
+		return ctrl.Result{}, err
 	}
 
-	if source.GetArtifact() == nil {
-		msg := "Source is not ready, artifact not found"
-		kustomization = kustomizev1.KustomizationNotReady(kustomization, "", kustomizev1.ArtifactFailedReason, msg)
-		if err := r.patchStatus(ctx, req, kustomization.Status); err != nil {
-			log.Error(err, "unable to update status for artifact not found")
-			return ctrl.Result{Requeue: true}, err
+	defer func() {
+		// record the value of the reconciliation request, if any
+		if v, ok := meta.ReconcileAnnotationValue(kustomization.GetAnnotations()); ok {
+			kustomization.Status.SetLastHandledReconcileRequest(v)
+		}
+
+		// Summarize Ready condition
+		conditions.SetSummary(kustomization,
+			meta.ReadyCondition,
+			conditions.WithConditions(
+				sourcev1.ArtifactAvailableCondition,
+				sourcev1.SourceAvailableCondition,
+			),
+		)
+
+		// Patch the object, ignoring conflicts on the conditions owned by
+		// this controller
+		patchOpts := []patch.Option{
+			patch.WithOwnedConditions{
+				Conditions: []string{
+					sourcev1.ArtifactAvailableCondition,
+					sourcev1.SourceAvailableCondition,
+					meta.ReadyCondition,
+					meta.ReconcilingCondition,
+					meta.StalledCondition,
+					meta.ProgressingReason,
```

Comment:
Why is here `ProgressingReason`? Do we need `HealtyCondition` also?

Reply:
I think we need to set `HealthyCondition` too